src/Cardano/WalletLayer/Wallet.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE TypeOperators #-}
+
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- | Module provides internal implementation of
+-- | wallet api
+
+module Cardano.WalletLayer.Wallet
+    (
+      CreateWallet
+    , WalletId
+    ) where
+
+import Prelude
+
+import Cardano.Wallet
+    ( Wallet )
+import Cardano.Wallet.AddressDerivation
+    ( Passphrase )
+import Cardano.Wallet.AddressDiscovery
+    ( AddressPoolGap )
+import Cardano.Wallet.Mnemonic
+    ( Mnemonic )
+import Data.Text
+    ( Text )
+import Data.Time.Units
+    ( Microsecond )
+import GHC.Generics
+    ( Generic )
+
+-- | Errors
+data CreateWalletError
+
+data GetWalletError
+
+-- | Types
+data CreateWallet = CreateWallet NewWallet | ImportWallet
+
+data NewWallet = NewWallet
+    {
+      _mnemonicSentence
+        :: !(Mnemonic 15)
+    , _mnemonicSentencePassphrase
+        :: !(Maybe (Passphrase "generation"))
+    , _name
+        :: !WalletName
+    , _passphrase
+        :: !(Passphrase "encryption")
+    , _addressPoolGap
+        :: !AddressPoolGap
+    } deriving (Show, Generic)
+
+newtype WalletName = WalletName { _getWalletName :: Text }
+    deriving (Eq, Show)
+
+newtype WalletId = WalletId Text
+    deriving (Eq, Show)
+
+newtype WalletTimestamp = WalletTimestamp Microsecond
+    deriving (Eq, Ord, Show)
+
+data WalletOperation = Ready | Restoring deriving (Eq, Show)
+
+data Delegation = Delegated | NotDelegated deriving (Eq, Show)
+
+newtype PassphraseInfo = PassphraseInfo { lastUpdated :: WalletTimestamp }
+    deriving (Eq, Show)
+
+data WalletMetadata = WalletMetadata {
+      id
+        :: !WalletId
+    , name
+        :: !WalletName
+    , addressPoolGap
+        :: !AddressPoolGap
+    , passphraseInfo
+        :: !PassphraseInfo
+    , state
+        :: !WalletOperation
+    , delegation
+        :: !Delegation
+    } deriving (Eq, Show, Generic)
+
+
+-- | Internal calls
+_createWallet
+    :: CreateWallet
+    -> m (Either CreateWalletError (Wallet WalletMetadata))
+_createWallet (CreateWallet (NewWallet _m1 _m2 _n _p _a)) = undefined
+_createWallet ImportWallet = undefined
+
+_getWallet
+    :: WalletId
+    -> m (Either GetWalletError (Wallet WalletMetadata))
+_getWallet (WalletId _id) = undefined
```

Comment:
I'll suggest to just leave this for later. We can have the 'WalletLayer' type / interface declared without any implementation for now :+1:

Reply:
removed